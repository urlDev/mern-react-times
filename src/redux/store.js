import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));