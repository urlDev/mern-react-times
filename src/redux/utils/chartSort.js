const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;