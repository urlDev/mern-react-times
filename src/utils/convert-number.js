const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
