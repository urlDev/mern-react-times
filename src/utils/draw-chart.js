// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
