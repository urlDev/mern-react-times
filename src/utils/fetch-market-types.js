import { changeMarketType, changeMarketName } from "../redux/actions/chart";

const fetchMarketTypes = (dispatch, market) => {
  //  One api endpoint can be used for all market quotes
  //  So I first determine the market type
  //  Then, checking which market type client has clicked on
  //  According to market, I pre-determined tickers/symbols
  //  With a switch function, I dispatch the related market types
  //  desired tickers to redux store.
  //  So with all this, I can use one fetch function for all market kinds,
  //  Therefore, less fetch calls, less code.
  //  More comments as it seemsüòÅ
  const Indexes = "%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE";
  const Crypto = "BTCUSD,LTCUSD,XLMUSD,BCNUSD,ETHUSD,ETCUSD";
  const Forex = "EURUSD,USDJPY,GBPUSD,EURGBP,EURJPY,GBPJPY";
  const Stocks = "AAPL,FB,GOOG,TSLA,NFLX,AMZN";
  const Commodities = "ZGUSD,CLUSD,HGUSD,SIUSD,PLUSD,BZUSD";

  switch (market) {
    case "Indexes":
      return [
        dispatch(changeMarketType(Indexes)),
        dispatch(changeMarketName("Indexes")),
      ];
    case "Crypto":
      return [
        dispatch(changeMarketType(Crypto)),
        dispatch(changeMarketName("Crypto")),
      ];
    case "Forex":
      return [
        dispatch(changeMarketType(Forex)),
        dispatch(changeMarketName("Forex")),
      ];
    case "Stocks":
      return [
        dispatch(changeMarketType(Stocks)),
        dispatch(changeMarketName("Stocks")),
      ];
    case "Commodities":
      return [
        dispatch(changeMarketType(Commodities)),
        dispatch(changeMarketName("Commodities")),
      ];
    default:
      return Indexes;
  }
};

export default fetchMarketTypes;
