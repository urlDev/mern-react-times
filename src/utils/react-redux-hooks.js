// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
