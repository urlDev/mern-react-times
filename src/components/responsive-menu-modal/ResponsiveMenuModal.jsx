import { useDispatch } from "../../utils/react-redux-hooks";

import {
  changeHeader,
  clearError,
  closeResponsiveMenu,
} from "../../redux/actions/news";
import { topics } from "../story-topic-headers/StoryTopicHeaders";

import {
  MenuModalAllPage,
  MenuModalContainer,
} from "./ResponsiveMenuModal.styles";

import { StyledLink } from "../story-topic-headers/StoryTopicHeaders.styles";

const ResponsiveMenuModal = () => {
  const dispatch = useDispatch();

  return (
    <MenuModalAllPage onClick={() => dispatch(closeResponsiveMenu())}>
      <MenuModalContainer>
        {topics.map((topic) => (
          <StyledLink key={topic} to={`/${topic.toLowerCase()}`}>
            <span
              onClick={() => {
                dispatch(changeHeader(topic.toLowerCase()));
                dispatch(closeResponsiveMenu());
                dispatch(clearError());
              }}
            >
              {topic}
            </span>
          </StyledLink>
        ))}
      </MenuModalContainer>
    </MenuModalAllPage>
  );
};

export default ResponsiveMenuModal;
