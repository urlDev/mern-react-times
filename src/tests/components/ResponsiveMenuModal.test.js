import { mount } from "enzyme";
import { BrowserRouter as Router } from "react-router-dom";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import {
  CHANGE_HEADER,
  CLEAR_ERROR,
  CLOSE_RESPONSIVE_MENU,
} from "redux/actions/news";

import ResponsiveMenuModal from "components/responsive-menu-modal/ResponsiveMenuModal";

describe("Testing ResponsiveMenuModal component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    store = mockStore({
      responsiveMenu: true,
    });

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    wrapper = mount(
      <Router>
        <ResponsiveMenuModal store={store} />
      </Router>
    );
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });
  test("Should render the component", () => {
    expect(wrapper).toHaveLength(1);
  });

  test("Should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Should close responsive menu on click container click", () => {
    wrapper.find("div").first().simulate("click");
    const actions = store.getActions();

    expect(actions).toEqual([{ type: CLOSE_RESPONSIVE_MENU }]);
  });

  test("Should close responsive menu on menu click", () => {
    wrapper.find("span").first().simulate("click");
    const actions = store.getActions();

    expect(actions).toEqual([
      { type: CHANGE_HEADER, payload: "home" },
      { type: CLOSE_RESPONSIVE_MENU },
      { type: CLEAR_ERROR },
      { type: CLOSE_RESPONSIVE_MENU },
    ]);
  });
});
