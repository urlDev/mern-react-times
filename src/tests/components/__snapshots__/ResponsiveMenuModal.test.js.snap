// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ResponsiveMenuModal component Should match the snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ResponsiveMenuModal
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <styled.div
        onClick={[Function]}
      >
        <div
          className="sc-jSgupP gylrLH"
          onClick={[Function]}
        >
          <styled.div>
            <div
              className="sc-gKsewC cPEtrP"
            >
              <Styled(Link)
                key="Home"
                to="/home"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/home"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/home"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/home"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Home
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Arts"
                to="/arts"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/arts"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/arts"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/arts"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Arts
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Business"
                to="/business"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/business"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/business"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/business"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Business
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Health"
                to="/health"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/health"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/health"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/health"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Health
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Opinion"
                to="/opinion"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/opinion"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/opinion"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/opinion"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Opinion
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Politics"
                to="/politics"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/politics"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/politics"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/politics"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Politics
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Science"
                to="/science"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/science"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/science"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/science"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Science
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Sports"
                to="/sports"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/sports"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/sports"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/sports"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Sports
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Technology"
                to="/technology"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/technology"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/technology"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/technology"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Technology
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <Styled(Link)
                key="Travel"
                to="/travel"
              >
                <Link
                  className="sc-eCssSg eJegWd"
                  to="/travel"
                >
                  <LinkAnchor
                    className="sc-eCssSg eJegWd"
                    href="/travel"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-eCssSg eJegWd"
                      href="/travel"
                      onClick={[Function]}
                    >
                      <span
                        onClick={[Function]}
                      >
                        Travel
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </ResponsiveMenuModal>
  </Router>
</BrowserRouter>
`;
